### PR TITLE
POC: embed weights as constants while tracing to specify what we want to save.

### PR DESCRIPTION
**Pull request by zdevito**
possibly easier approach than other ways of deciding what to save.

### CODE REVIEW

**Comment by apaszke** [comment]
That will break autograd closure generation once it lands. We're saving constants as tensors, and wrapping them in Variables every time they're used. However, if you want user-visible Parameter objects to receive gradients, you need to preserve Variable identity

**Comment by zdevito** [comment]
This pathway is only intended to be used as way to embed parameters into exported ToffeeIR, not as part of the JIT. Currently we have a bunch of side-channel/positional hacks to specify what we want to export in the graph. It seems much simpler to just specify this information while you record the trace, which is what this commit enables.

**Comment by apaszke** [comment]
Except that some constants caught in the tracer might not be model parameters (e.g. the one that appears in `addmm`)

**Comment by ezyang** [comment]
Yeah, I guess the biggest problem with this patch is that it collapses the distinction between model parameter and constant. So, it's OK if you can truly treat these equivalently, but if you want to e.g. train the models this is likely to cause problems.

Maybe the positional passing fiasco will get better when we start actually using state_dict() and start setting debug names for parameters?

**Comment by apaszke** [comment]
I'd go with the way Ed's patch handles buffers

**Comment by zdevito** [comment]
We need to talk in person about this. This patch removes code and solves a problem we have with exporting models. It is orthogonal to batchnorm buffers. So we need to figure out where the misunderstanding is.
